akka-stream-tests/src/test/scala/akka/stream/impl/fusing/InterpreterSpec.scala

```diff
@@ -569,39 +569,64 @@ class InterpreterSpec extends AkkaSpec with GraphInterpreterSpecKit {
 
   }
 
-  private[akka] case class Doubler[T]() extends PushPullStage[T, T] {
-    var oneMore: Boolean = false
-    var lastElem: T = _
-
-    override def onPush(elem: T, ctx: Context[T]): SyncDirective = {
-      lastElem = elem
-      oneMore = true
-      ctx.push(elem)
+  private[akka] case class Doubler[T]() extends GraphStage[FlowShape[T, T]] {
+    val out: Outlet[T] = Outlet("Doubler-out")
+    val in: Inlet[T] = Inlet("Doubler-in")
+
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+      var latest: T = _
+      var oneMore: Boolean = false
+      setHandler(in, new InHandler {
+        override def onPush(): Unit = {
+          latest = grab(in)
+          oneMore = true
+          push(out, latest)
+        }
+      })
+
+      setHandler(out, new OutHandler {
+        override def onPull(): Unit = {
+          if (oneMore) {
+            oneMore = false
+            push(out, latest)
+          } else {
+            pull(in)
+          }
+        }
+      })
     }
 
-    override def onPull(ctx: Context[T]): SyncDirective = {
-      if (oneMore) {
-        oneMore = false
-        ctx.push(lastElem)
-      } else ctx.pull()
-    }
+    override def shape: FlowShape[T, T] = FlowShape(in, out)
   }
 
-  private[akka] case class KeepGoing[T]() extends PushPullStage[T, T] {
+  private[akka] case class KeepGoing[T]() extends GraphStage[FlowShape[T, T]] {
```

Comment:
See if you can apply some of the comments I made on Doubler on this conversion as well!